pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,394 @@
+package glusterfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/gluster/gluster-csi-driver/pkg/glusterfs/utils"
+
+	csi "github.com/container-storage-interface/spec/lib/go/csi/v0"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/golang/glog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	glusterDescAnn            = "GlusterFS-CSI"
+	glusterDescAnnValue       = "gluster.org/glusterfs-csi"
+	defaultVolumeSize   int64 = 1000 * utils.MB // default volume size ie 1 GB
+	defaultReplicaCount       = 3
+)
+
+var errVolumeNotFound = errors.New("volume not found")
+
+type ControllerServer struct {
+	*GfDriver
+}
+
+//CreateVolume creates and starts the volume
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	var glusterServer string
+	var bkpServers []string
+
+	if req == nil {
+		glog.Errorf("volume create request is nil")
+		return nil, status.Errorf(codes.InvalidArgument, "request cannot be empty")
+	}
+
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "Name is a required field")
+	}
+	glog.V(1).Infof("creating volume with name ", req.Name)
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities is a required field")
+	}
+
+	// If capacity mentioned, pick that or use default size 1 GB
+	volSizeBytes := defaultVolumeSize
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int64(req.GetCapacityRange().GetRequiredBytes())
+	}
+
+	volSizeMB := int(utils.RoundUpSize(volSizeBytes, 1024*1024))
+
+	// Get Volume name : TODO use the values from request
+	volumeName := req.Name
+	glusterVol := req.GetParameters()["glustervol"]
+	glusterServer = req.GetParameters()["glusterserver"]
+	glusterURL := req.GetParameters()["glusterurl"]
+	glusterURLPort := req.GetParameters()["glusterurlport"]
+	glusterUser := req.GetParameters()["glusteruser"]
+	glusterUserSecret := req.GetParameters()["glusterusersecret"]
+
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if len(volumeListResp) > 0 {
+		glusterServer, bkpServers, err := cs.checkExistingVolumes(&volumeListResp, volumeName, volSizeMB)
+		if err != nil && err != errVolumeNotFound {
+			return nil, err
+
+		}
+		if err == errVolumeNotFound {
+			goto VOLUMECREATE
```

Comment:
wow, Golang supports `goto`! This is pretty horrible.

Isnt `break` sufficient?

Reply:
yeah, the only real reasons for using goto, imho is common exit cleanup path. that's nicely taken care of by defer in golang....

Reply:
cant use `break` here, as there is no `for loop`.  to avoid code duplication I am using `GOTO`,
any other suggestion over here.

Reply:
You can use Volume Get by name and compare instead of list.